components/raftstore/src/store/fsm/peer.rs

```diff
@@ -2341,24 +2332,39 @@ where
                 "err" => %e,
             );
         }
-        let is_initialized = self.fsm.peer.is_initialized();
-        if let Err(e) = self.fsm.peer.destroy(
-            &self.ctx.engines,
-            &mut self.ctx.perf_context,
+        self.fsm.peer.destroy_all_pending_requests();
+
+        let store = self.fsm.peer.get_store();
+        let task = RegionTask::ClearMeta {
+            peer_id: self.fsm.peer_id(),
+            tag: self.fsm.peer.tag.clone(),
+            region: store.region().clone(),
+            raft_state: store.raft_local_state().clone(),
+            merge_state: self.fsm.peer.merge_state().cloned(),
+            notifier: Box::new(self.ctx.router.clone()),
             merged_by_target,
-        ) {
-            // If not panic here, the peer will be recreated in the next restart,
-            // then it will be gc again. But if some overlap region is created
-            // before restarting, the gc action will delete the overlap region's
-            // data too.
-            panic!("{} destroy err {:?}", self.fsm.peer.tag, e);
+        };
+
+        if let Err(e) = self.ctx.region_scheduler.must_schedule(task) {
+            panic!(
+                "failed to schedule clear meta task, region_id={}, peer_id={}, err={}",
+                self.fsm.region_id(),
+                self.fsm.peer_id(),
+                e
+            );
```

Comment:
I worry about the change that some metadata are changed before clearing meta and the others are changed after clearing meta.
You may take some time to figure out the correctness dependency of atomicity changing, which is the most difficult one in this PR.
My suggestion is to write kv db/raft db first, then change the memory state(like `StoreMeta`). I think it may introduce fewer bugs than the method used here.
However, it's still needed to check whether some other places read the `kv db` or `raft db` and the correctness isn't broken. (E.g. the create peer process will check the `RegionLocalState` in kv db)